driver/src/order_driver.rs

```diff
@@ -1,108 +1,170 @@
-            let mut state = db.get_balances_for_state_root(&state_root)?;
-
-            let mut orders = db.get_orders_of_slot(&slot)?;
-            let non_reserved_orders_hash = orders.rolling_hash(0);
-            hash_consistency_check(
-                non_reserved_orders_hash,
-                non_reserved_orders_hash_from_contract,
-                "non-reserved-orders",
-            )?;
-
-            let standing_orders = db.get_standing_orders_of_slot(&slot)?;
-
-            orders.extend(
-                standing_orders
-                    .iter()
-                    .filter(|standing_order| standing_order.num_orders() > 0)
-                    .flat_map(|standing_order| standing_order.get_orders().clone()),
-            );
-            info!("All Orders: {:?}", orders);
-
-            let standing_order_indexes = batch_index_from_standing_orders(&standing_orders);
-            let total_order_hash_from_contract =
-                contract.calculate_order_hash(slot, standing_order_indexes.clone())?;
-            let total_order_hash_calculated =
-                standing_orders.concatenating_hash(non_reserved_orders_hash);
-            hash_consistency_check(
-                total_order_hash_calculated,
-                total_order_hash_from_contract,
-                "overall-order",
-            )?;
-
-            let solution = if !orders.is_empty() {
-                price_finder
-                    .find_prices(&orders, &state)
-                    .unwrap_or_else(|e| {
-                        error!(
-                            "Error computing result: {}\n Falling back to trivial solution",
-                            e
-                        );
-                        Solution::trivial(orders.len())
-                    })
-            } else {
-                warn!("No orders in batch. Falling back to trivial solution");
-                Solution::trivial(orders.len())
-            };
-
-            // Compute updated balances
-            update_balances(&mut state, &orders, &solution);
-            let new_state_root = state.rolling_hash(state.state_index.low_u32() + 1);
-
-            info!(
-                "New AccountState hash is {}, Solution: {:?}",
-                new_state_root, solution
-            );
-
-            contract.apply_auction(
-                slot,
-                state_root,
-                new_state_root,
-                total_order_hash_from_contract,
-                standing_order_indexes,
-                solution.bytes(),
-            )?;
-            return Ok(true);
-        } else {
-            info!("Need to wait before processing auction slot {:?}", slot);
+use web3::types::{H256, U128, U256};
+
+use std::collections::HashMap;
+
+struct AuctionBid {
+    previous_state: H256,
+    new_state: H256,
+    solution: Solution,
+}
+
+pub struct OrderProcessor<'a, D: DbInterface, C: SnappContract> {
```

Comment:
Whenever we instantiate an order processor (e.g. using OrderProcessor::new) the lifetime of this object is implicitly defined by the lifetime of db, contract and price finder that we are passing along.

Since all of these are created in the same scope (cf. main.rs), the compiler can figure out that they anyways have the same lifetime and thus no explicit lifetime specification is needed. Let's chat offline to clarify more if needed.

Reply:
This is some really advance stuff. I'll get there at some point when I get some time to resume the book!